### ISSUE TITLE
Remove records for  IsCommutative<Foo>, Commutative<Foo> ?

### ISSUE THREAD

**Issue by mechvel** [agda/agda-stdlib#1162]
lib-1.3 has 10 parasitic records of (Is)Commutative<Foo>,
10 levels in the bundle hierarchy for algebra.
Code and effort are spent in filling instances of these records, in standard library as well as in applied programs.
 
This can be improved by defining Commutative<Foo> via  Σ:
```
module _ {α α=}
  where 
  CommutativeSemigroup :  Set _
  CommutativeSemigroup =  
           Σ (Semigroup α α=) (\H → let open Semigroup H in Commutative _≈_ _∙_) 
  ...
  CommutativeRing :  Set _
  CommutativeRing =  
                  Σ (Ring α α=) (\R → let open Ring R in Commutative _≈_ _*_) 
```
These are five declarations.
Usage:
```
foo : {α α=} → CommutativeRing α α= → Foo
foo (R , *-comm) =
                 < evaluate with using (R : Ring) and *-comm >
```
For example, this allows us to remove 13 record instances of IsCommutative and Commutative from Data.Integer.Properties.
  
A small drawback: for example,
instead of  ``CommutativeRing.1#  C`` one will need to write ``Ring.1# (proj₁ C)``, 
or  ``let (R, *comm) = C in Ring.1# R``
(and ``*comm`` will be sometimes used ...).
Similarly it is with  ``open Commutative<Foo>``
(if only Σ does not export from its first argument record).

?

**Comment by MatthewDaggitt** [comment]
My question is why would you single out commutativity? Why not associativity, idempotence, inverses, identities as well? There's nothing special about commutativity. But at that point the question is why would we have bundles at all?

Note that you can already use the style you propose above:
```agda
module {α α=} (S : Semiring α α=)
       (open Semiring S)
       (comm : Commutative _+_)
       where
```

**Comment by mechvel** [comment]
> My question is why would you single out commutativity? Why not associativity, idempotence, 
> inverses, identities as well?

Compare the signature decls:
``(I)     (S : CommutativeSemiring _ _) → Foo``
to
``(II)    (S : Semiring _ _) → IsCommutative S → Foo``

II  is only 7 characters loger.
But the approach of II saves two records in the library:
                    IsCommutativeSemiring and CommutativeSemiring
- two levels in the hierarchy.
These records have several fields to be filled with instances each time.

Concerning associativity
--------------------------------

If it was  AssociativeMagma  instead of  Semigroup,
then one could think of separating the property and declaring
 ``  (M : Magma _ _) → IsAssociative M → Foo``

But the classical tradition is to call it  Semigroup.
And changing "R : CommutativeRing"  to  ``(R : Ring) → IsCommutative R``
does not deviate from the tradition.

Also almost all classical algebraic class instances (the ones considered in classical textbooks) are associative:
Semigroup, Monoid, Group, Semiring, Ring, Field ...
"Almost everything interesting" in algebra inherits some associative operation and may be another associative operation. This is because without associativity very few interesting properties can
be proved. As exclusion we have, say, Lie algebra, but this is far from standard library.

``Commutative`` is different. About half of the instances (considered in textbooks) of the above classes are not commutative. A semigroup is often commutative and also is often non-commutative,
the same is with Monoid, Group, Ring.
Also for this reason, it has sense to write ``(H : Semigroup) →``
rather than ``(H : Magma) → IsAssociative H →``.

This is why i think that "commutative" can be treated differently from "associative".

Concerning idempotence, inverses, identities and such
-------------------------------------------------------------------------

One can write instead of  ``(G : Group _ _) → Foo``
```
  (G : Semigroup _ _)
  (open Semigroup G) (ε : Carrier) (_⁻¹ : Op₁ Carrier)
  (ε∙ : LeftIdentity _≈_ ε _∙_) (∙ε : RinghIdentity _≈_ ε _∙_) 
  (invLaws : Inverse _≈_ ε _∙_ _⁻¹)
```
But a notion of Group has been introduced in order of not to write such a large expression many times, because this notion is often used.
For the same reason the Agda library declares this notion as a record.

This also responds to the question
> But at that point the question is why would we have bundles at all?

As to idempotence,  standard library has not any corresponding domain class. Has it, say, IdempotentSemigroup ?
Concerning identities:  to define an identity law field, one needs to first introduce ``ε`` - a non-property operation. So that we need to add two items. Two items deserve a record level
more than a single item Commutative. Similar is with inverses.

> Note that you can already use the style you propose above:
``` 
> module {α α=} (S : Semiring α α=)
>        (open Semiring S)
>        (comm : Commutative _+_)
>        where
```
1) This is too lengthy. My idea was to skip ``open``.
2) My idea was also to save _several record levels_ in Standard library. Because the word    
  Commutative intrudes in several places there, and each such place spends two additional
  record declarations. 
3) Suppose the programmer uses some ``lemma`` from standard library, and this lemma  has 
    ``H : CommutativeSemigroup``  as argument  (there are such lemmas in lib-1.3).
Then the programmer is forced to fill with the instances the field records of 
IsCommutativeSemigroup and CommutativeSemigroup.
4) I hope standard library will add some algebraic instances for ``_×_, Vector, Fraction``. 
They will deal with commutativity.

I set forth the idea so profoundly that I start to fear of standard library to take the approach.

**Comment by mechvel** [comment]
Summary
-------------
We have 10 extra records and 10 extra levels in the algebraic bundle hierarchy: 
IsCommutative<Foo> and Commutative<Foo> records for 
Semigroup, Monoid, Group, Semiring, Ring.

This implies the need to fill in the instances for several their fields, to do this in the standard library code and also in applied programs.
This all can be replaced with setting ``(A : Foo _ _) → IsCommutative A →`` 
instead of  ``(A : Commutative<Foo> _ _) → ``,
and making IsCommutative a _predicate_ over domains.

**Comment by gallais** [comment]
Are you suggesting we should have a different `IsCommutative` predicate for
each possible `Foo`? How is this more convenient that having well defined
bundles?

**Comment by mechvel** [comment]
There are five definitions:
```
IsCommutativeSemigroup : ∀ {α α=} → Semigroup α α= → Set _
IsCommutativeSemigroup S =   Commutative _≈_ _∙_  where open Semigroup S
...
IsCommutativeRing : ∀ {α α=} → Ring α α= → Set _
IsCommutativeRing R =  Commutative _≈_ _*_  where open Ring R
```
Example of usage:
```
foo : ∀ {α α=} → (R : Ring α α=) → IsCommutativeRing R → Ring.Carrier R
foo R _ =  1# * 1#
                where open Ring R
fooℤ : ℤ
fooℤ = foo ℤp.+-*-ring ℤp.*-comm  
```
And no need in keeping 10 record bundles in the hierarchy and in filling in several fields in each of them.

**Comment by mechvel** [comment]
All right, this is possible:
```
module _ {α α=}
  where 
  CommutativeSemigroup :  Set _
  ...

  CommutativeRing :  Set _
  CommutativeRing =  Σ (Ring α α=) (\R → let open Ring R in Commutative _≈_ _*_) 
```
Usage:
```
foo : {α α=} → CommutativeRing α α= → Foo
foo (R , *-comm) =
                 < evaluate with using R and *-comm >
```
For example, this allows us to remove 13 records of IsCommutative<Foo> and Commutative<Foo> from Data.Integer.Properties.

**Comment by mechvel** [comment]
The proposal has totally changed.
See the first message in the issue.

**Comment by mechvel** [comment]
> My question is why would you single out commutativity? Why not associativity, idempotence, 
> inverses, identities as well? There's nothing special about commutativity. But at that point the 
> question is why would we have bundles at all?

(This was asked about the deleted proposal version. But this also relates to this version).

1) "Commutative" takes place in 10 bundles, while "Associative" is only in one bundle, the effect
     is much smaller. 
2) Replacing other bundles with \Sigma will make it difficult to extract operations from bundles:
    complex patterns, many composed ``proj`` to apply.

**Comment by gallais** [comment]
> 10 levels in the bundle hierarchy for algebra.
> (...)
> These are five declarations.

Well, sure: if you remove features there are fewer of them in the source code.
However I cannot express `IsCommutativeRing 0 1 _+_ _*_` (or whatever the syntax
is) conveniently anymore. I have to roll out my own `Σ`-bundled `IsRing` & the
commutative constraint. And when I want to use a `CommutativeX`, I cannot simply
use the one declared in the library, I need to explicitly pair the proof of `X`
and that of `Commutative X-op`.

I really don't see how any of this is a win.

**Comment by mechvel** [comment]
> However I cannot express IsCommutativeRing 0 1 _+_ _*_ (or whatever the syntax is) 
> conveniently anymore.
 
IsCommutativeRing  will not be needed.  IsRing will be sufficient.
I do not catch your idea.
Please, give  a concrete example: some piece of code in terms of lib-1.3, 
and I shall try to show its replacement in the proposed library.

**Comment by MatthewDaggitt** [comment]
@mechvel Firstly if you want to change your suggestion, please do not edit previous comments as it means that the conversation no longer makes sense to people reading the conversation for the first time. Instead add your new proposal as a new comment at the bottom. You should only edit your previous comments in order to fix typos or improve clarity.

Secondly I'm pleased to find that you're gradually coming around to the way the library works at the moment! Your new suggestion:
```agda
CommutativeSemigroup :  Set _
CommutativeSemigroup =  
           Σ (Semigroup α α=) (\H → let open Semigroup H in Commutative _≈_ _∙_) 
```
of having a sigma type with a bundle and a property is almost exactly what we already have:
```agda
record CommutativeSemigroup : Set ...
  field
    semigroup   : Semigroup ...
    commutative : Commutative ...
```
but with less features! e.g. the ability to add extra declarations within the record, or to open the semigroup bundle.

Thirdly, I don't see how your new proposal solves your stated aim of minimising the number of declarations you have to make when adding a new datatype. You're still going to want to define things as like:
```agda
+-commutativeSemigroup : CommutativeSemigroup
+-commutativeSemigroup = +-semigroup, +-commutative
``` 

Fourthly, as I said above, the addition of a short single line allows you to use the style of module arguments that you desire. I know you said adding the `open` line in
```agda
module {α α=} (S : Semiring α α=)
       (open Semiring S)
       (comm : Commutative _+_)
       where
```
was too long, but I would respectively disagree. It's a single, short line that's needed at most once per module.

Fifthly, this is a huge breaking change to the library and even your claimed benefit, which even if  @gallais and I agreed existed, is fairly small one.

A lot of thought has gone into the record hierarchies in the library and we're relatively certain that, at least under the current constraints of the Agda language, this is the design that best accommodates the many different ways people need to use the hierarchies. (@JacquesCarette and others are researching better language designs to represent and construct such hierarchies, but these are not available yet.)

In summary we won't be making this change, apologies. Thank you for the suggestions and the discussion though!

**Comment by mechvel** [comment]
>  I'm pleased to find that you're gradually coming around to the way the
> library works at the moment! Your new suggestion:
```
  CommutativeSemigroup :  Set _
  CommutativeSemigroup =  
      Σ (Semigroup α α=) (\H → let open Semigroup H in Commutative _≈_ _∙_) 
```
>  of having a sigma type with a bundle and a property is almost exactly what
 > we already have:
```
  record CommutativeSemigroup : Set ...
    field
      semigroup   : Semigroup ...
      commutative : Commutative ...
```

No, lib-1.3 has
```
record CommutativeSemigroup c ℓ : Set (suc (c ⊔ ℓ)) where                         
  ...
  field                                                                           
    Carrier                : Set c                                               
    _≈_                    : Rel Carrier ℓ                                       
    _∙_                    : Op₂ Carrier                                         
    isCommutativeSemigroup : IsCommutativeSemigroup _≈_ _∙_
    ...
    semigroup = ...
```  
In particular it also applies declaring and using record  IsCommutativeSemigroup,
which is not needed to introduce in my proposal.

> but with less features! e.g. the ability to add extra declarations within
> the record, or to open the semigroup bundle.

For "Commutative" there are not needed any extra declarations inside.
For example, ``semigroup`` is already in ``Σ``.

> Thirdly, I don't see how your new proposal solves your stated aim of
> minimising the number of declarations you have to make when adding a new
> datatype. You're still going to want to define things as like:
>
> +-commutativeSemigroup : CommutativeSemigroup
> +-commutativeSemigroup = +-semigroup, +-commutative

I have somehow missed this point. 
All right, but it saves the ``IsCommutative<Foo>`` records. There are five of them.
By the way, can the field
```
  isCommutativeSemigroup : IsCommutativeSemigroup _≈_ _∙_
```
in lib-1.3 be changed to
```
  ∙-comm : Commutative _≈_ _∙_
```
Will it become simpler and better?
